gemini/use-cases/retrieval-augmented-generation/code_rag.ipynb

```diff
@@ -0,0 +1,1231 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #35.            while cursor.goto_first_child():
</pre><p>This should probably be simplified to reduce duplicated code</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/481/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Reply:
<p>This is the intended way to use it. During syntax tree parsing the cursor needs to visit all the nodes of the syntax tree. The tree object has a different walk mechanism which requires the cursor to move "up", "down", "left" and "right" to identify the target node level at which we want to extract code snippets. This double while loop setup does that effectively. Although it still looks complex but it works better and intuitive to parse only intended nodes.</p>

Reply:
<p>Ok</p>